src/type-resolution/resolve.ts

```diff
@@ -0,0 +1,167 @@
+import * as reflect from 'jsii-reflect';
+import {
+  isBehavioralInterface,
+  isEnum,
+  isEnumLikeClass,
+  isSerializableInterface,
+} from '../jsii2schema';
+import { TemplateExpression } from '../parser/template';
+import { resolveInitializerExpression } from './callables';
+import {
+  assertArrayOfType,
+  assertMapOfType,
+  resolveArrayOfTypeExpression,
+  resolveMapOfTypeExpression,
+} from './collections';
+import {
+  assertClass,
+  assertEnum,
+  resolveEnumExpression,
+  resolveEnumLikeExpression,
+} from './enums';
+import { assertExpressionType, TypedTemplateExpression } from './expression';
+import { resolveDateLiteral } from './literals';
+import {
+  assertLiteralOrIntrinsic,
+  assertVoid,
+  isAny,
+  resolveAnyExpression,
+  resolveVoidExpression,
+} from './primitives';
+import { assertRef, refOrResolve } from './references';
+import { isConstruct } from './resource-like';
+import { assertInterface, resolveStructExpression } from './struct';
+import { assertUnionOfTypes, resolveUnionOfTypesExpression } from './union';
+
+export function resolveExpressionType(
+  x: TemplateExpression,
+  typeRef: reflect.TypeReference
+): TypedTemplateExpression {
+  const expectedType = analyzeTypeReference(typeRef);
+
+  switch (expectedType) {
+    case ResolvableExpressionType.NUMBER:
+      return assertLiteralOrIntrinsic(x, 'number');
+    case ResolvableExpressionType.BOOLEAN:
+      return assertLiteralOrIntrinsic(x, 'boolean');
+    case ResolvableExpressionType.STRING:
+      return assertLiteralOrIntrinsic(x, 'string');
+    case ResolvableExpressionType.DATE:
+      return resolveDateLiteral(assertExpressionType(x, 'string'));
+    case ResolvableExpressionType.ARRAY_OF_TYPE:
+      return resolveArrayOfTypeExpression(
+        assertExpressionType(x, 'array'),
+        assertArrayOfType(typeRef)
+      );
+    case ResolvableExpressionType.MAP_OF_TYPE:
```

Comment:
Something just popped into my head. I wonder if we need an additional syntax element to specify maps where the keys are also specified as `rvalues` (for lack of a better term).

In code, you can do:

```ts
functionCall({
  [bucket.bucketName]: 'foo'
})
```

But in JSON/YAML you cannot do:

```yaml
{
  FunctionCall: {
    { 'Fn::GetProp': [Bucket, BucketName] }: foo
  }
}
```

We would need to support additional syntax like:

```yaml
{
  FunctionCall: {
    'Fn::Map': [
      { Key: { 'Fn::GetProp': [Bucket, BucketName] }, Value: 'foo' },
    ]
  }
}
```